### ISSUE TITLE
Loss estimation is very slow

### ISSUE THREAD

**Issue by sudarshanv01** [sudarshanv01/coeffnet#23]
The current way of estimating loss is done graph-by-graph and is very slow. 

```python
def coeff2density_loss(data, predicted_y, do_backward=True):
    """Get the loss when converting the coefficient matrix to the density."""

    real_y = data.x_transition_state

    for i in range(data.num_graphs):
        real_c_ij = real_y[data.batch == i]
        predicted_c_ij = predicted_y[data.batch == i]

        loss_positive = F.mse_loss(predicted_c_ij, real_c_ij, reduction="sum")
        loss_negative = F.mse_loss(-predicted_c_ij, real_c_ij, reduction="sum")

        if loss_positive.item() < loss_negative.item():
            loss = loss_positive
        else:
            loss = loss_negative

        if do_backward:
            loss.backward(retain_graph=True)

        yield loss.item()
```

What would improve the speed would be to generate an `itertools.product` of all possible "phases" (either 1 or -1) to multiply the prediction with so that it matches that of the computed eigenvectors. That would generate a few `loss` functions, but would allow for everything to still be a `tensor`. Potentially create a custom loss function to achieve this.
